public/externalLibs/sound/sounds.js

```diff
@@ -111,33 +111,139 @@ function get_duration(sourcesound) {
   return tail(sourcesound)
 }
 
-function is_sound(sound) {
-  return is_pair(sound) && head(sound) === 'sound'
-}
-
-var _playing = false
-var _audio = null
-function play(sound) {
-  if (_playing) return
-
-  if (!is_sound(sound)) {
-    throw new Error('play() expects sound as input, did you forget to sourcesound_to_sound()?')
-  }
-
-  var data = tail(sound)
+var _playing = false;
+var player;
+
+function play(sourcesound) {
+    // If a sourcesound is already playing, terminate execution
+    if (_playing) return;
+    _playing = true;
+
+    // Declaring duration and wave variables
+    var duration = get_duration(sourcesound);
+    var wave = get_wave(sourcesound);
+
+    // Create AudioContext
+    //const AudioContext = window.AudioContext || window.webkitAudioContext;
+    
+    // Main audio context
+    player = new AudioContext();
+
+    // Controls Length of buffer in seconds.
+    var buffer_length = 0.1;
+
+    // Define Buffer Size
+    var bufferSize = FS * buffer_length;
+
+    // Create two buffers
+    var buffer1 = player.createBuffer(1, bufferSize, FS);
+    var buffer2 = player.createBuffer(1, bufferSize, FS);
+
+    // Create Discretized Data Set
+    let data1 = buffer1.getChannelData(0);
+    let data2 = buffer2.getChannelData(0);
+
+    // Discretizes a sourcesound to a sound starting from elapsed_duration, for length seconds
+    function discretize_from(wave, duration, elapsed_duration, length, data) {
+        if (elapsed_duration + length > duration) {
+            for (var i = elapsed_duration * FS; i < duration * FS; i++) {
+                data[i - elapsed_duration * FS] = wave(i / FS);
+            }
+            return data;
+        } else if (duration - elapsed_duration > 0) {
+            for (var i = elapsed_duration * FS; i < (elapsed_duration + length) * FS; i++) {
+                data[i - elapsed_duration * FS] = wave(i / FS);
+            }
+            return data;
+        }
+    }
 
-  _audio = raw_to_audio(data)
-  _audio.addEventListener('ended', stop)
-  _audio.play()
-  _playing = true
+    // Schedules playback of sounds (Once keeps track of whether first instance, first is first or second sound)
+    function ping_pong(wave, duration, elapsed_duration,
```

Comment:
This seems to be core of your new implementation. Could you elaborate more on what this function actually does? 

It is not super clear from your comments.

Reply:
Hi! I've refactored my play() function in the latest commit. Removed loads of repetitive code. Let me know if its still unclear